### PR TITLE
focus: beyond fullscreen when focused explicitly

### PR DESCRIPTION
**Pull request by lbonn**
When issuing a focus command on a specific container, users expect to proceed it even if is hidden by a fullscreen window.

This matches the behavior of i3.

Reported by @markstos: https://github.com/lbonn/i3-focus-last/issues/5

### CODE REVIEW

File: sway/input/seat.c
Comment (Xyene) [inline sway/input/seat.c:1062]: This comment probably belongs in the header file, distinguishing `seat_set_focus` from `seat_set_focus_force`.

File: sway/input/seat.c
Comment (Xyene) [inline sway/input/seat.c:1061]: Nit: we usually shy away from `_function` naming internal functions.

File: sway/input/seat.c
Comment (Xyene) [inline sway/input/seat.c:1096]: Design question: could we refactor out this code into a new `struct sway_container *container_obstructing_fullscreen_container(struct sway_container *)`?

Then `focus.c` could check if `container_obstructing_fullscreen_container(...) != NULL` and `container_fullscreen_disable(...)` that, removing the need to plumb `force` through this code.

**Review by Xyene — COMMENTED** [review]
Sorry for the late review; I have some general questions regarding this approach, but functionally it looks good.

File: sway/input/seat.c
Comment (lbonn) [reply sway/input/seat.c]: This sounds like a much better design, thank you.
I've made the changes accordingly.

**Comment by lbonn** [comment]
After @Xyene's review, the method has changed a bit.

However the behavior is different than last time I've tried and fullscreen windows unfocused this way will still appear maximized if focused again. I'll try to see if it comes from my patch or recent changes since the PR was opened.

**Comment by lbonn** [comment]
Actually it also happens on the original version... Any idea what might be going on?

**Comment by Xyene** [comment]
Hmm, I'm not sure, but you might be missing a call to `arrange_root` after the disable?

File: sway/input/seat.c
Comment (Xyene) [reply sway/input/seat.c]: :+1: Ideally, we'd use this code within `seat_set_focus` too, so we don't have the same block duplicated twice.

File: sway/commands/focus.c
Comment (Xyene) [inline sway/commands/focus.c:380]: Probably best to add a comment here for why `seat_set_focus_container` is insufficient, since this is the only special case.

File: sway/input/seat.c
Comment (lbonn) [reply sway/input/seat.c]: Which block are you talking about? The one added in focus.c?

Initially, that's why I added a `seat_set_focus_force` function as when I reviewed the code back in January it seemed like this was the only location we'd actually need to do that. There are a lot of calls to `seat_set_focus*` and adding the same logic there has a big chance of introducing regressions IMO.

**Comment by lbonn** [comment]
> Hmm, I'm not sure, but you might be missing a call to `arrange_root` after the disable?

That did the trick, thank you! Just pushed a new version with `arrange_root()` and an explanatory comment.

File: sway/input/seat.c
Comment (Xyene) [reply sway/input/seat.c]: I meant that these two blocks are basically identical in functionality:

https://github.com/swaywm/sway/blob/63564ee11dd168e35b7e1ac7f3068fff10372cbd/sway/tree/container.c#L420-L441

https://github.com/swaywm/sway/blob/63564ee11dd168e35b7e1ac7f3068fff10372cbd/sway/input/seat.c#L1093-L1112

The one from `seat.c` can be made to call `container_obstructing_fullscreen_container` and avoid duplicating its logic.

Comment (lbonn) [reply sway/input/seat.c]: Ah gotcha haha :smiley: !

File: sway/tree/container.c
Comment (Xyene) [inline sway/tree/container.c:423]: Most of the functions here don't do nullchecks internally, let's just move this check to the caller for consistency.

**Review by Xyene — COMMENTED** [review]
Otherwise LGTM.

File: sway/tree/container.c
Comment (lbonn) [reply sway/tree/container.c]: Updated

**Comment by markstos** [comment]
@emersion This bug fix to match i3 behavior appears ready to merge pending final review by @Xyene and could be a good candidate for the 1.6 release.

**Review by Xyene — APPROVED** [review]
Thanks!

**Pull request merged**
